boot/zephyr/serial_adapter.c

```diff
@@ -53,138 +53,154 @@ static int boot_uart_fifo_init(void);
-     * used_queue.
-     */
-    avail_queue = &free_queue;
-    lines_queue = &used_queue;
+	for (i = 0; i < ARRAY_SIZE(line_bufs); i++) {
+		sys_slist_append(&free_queue, &line_bufs[i].node);
+	}
 
-    return boot_uart_fifo_init();
+	/* Zephyr UART handler takes an empty buffer from free_queue,
+	 * stores UART input in it until EOL, and then puts it into
+	 * used_queue.
+	 */
+	avail_queue = &free_queue;
+	lines_queue = &used_queue;
+
+	return boot_uart_fifo_init();
 }
 
 static void
 boot_uart_fifo_callback(struct device *dev)
 {
-    static struct line_input *cmd;
-    u8_t byte;
-    int rx;
-
-    while (uart_irq_update(uart_dev) &&
-           uart_irq_rx_ready(uart_dev)) {
-
-        rx = uart_fifo_read(uart_dev, &byte, 1);
-
-        if (!cmd) {
-            cmd = k_fifo_get(avail_queue, K_NO_WAIT);
-            if (!cmd) {
-                return;
-            }
-        }
-
-        if (cur < CONFIG_BOOT_MAX_LINE_INPUT_LEN) {
-            cmd->line[cur++] = byte;
-        }
-
-        if (byte ==  '\n') {
-            cmd->len = cur;
-            k_fifo_put(lines_queue, cmd);
-            cur = 0;
-        }
-
-    }
+	static struct line_input *cmd;
+	u8_t byte;
+	int rx;
+
+	while (uart_irq_update(uart_dev) &&
+	       uart_irq_rx_ready(uart_dev)) {
+
+		rx = uart_fifo_read(uart_dev, &byte, 1);
+		if (rx != 1) {
+			continue;
+		}
+
+		if (!cmd) {
+			sys_snode_t *node;
+
+			node = sys_slist_get(avail_queue);
+			if (!node) {
+				return;
+			}
+			cmd = CONTAINER_OF(node, struct line_input, node);
+		}
+
+		if (cur < CONFIG_BOOT_MAX_LINE_INPUT_LEN) {
+			cmd->line[cur++] = byte;
+		}
+
+		if (byte ==  '\n') {
+			cmd->len = cur;
+			sys_slist_append(lines_queue, &cmd->node);
+			cur = 0;
+		}
+	}
 }
 
 static int
 boot_uart_fifo_getline(char **line)
 {
-    static struct line_input *cmd;
+	static struct line_input *cmd;
+	sys_snode_t *node;
 
-    /* Recycle cmd buffer returned previous time */
-    if (cmd != NULL) {
-        k_fifo_put(&free_queue, cmd);
-    }
+	/* Recycle cmd buffer returned previous time */
+	if (cmd != NULL) {
+		sys_slist_append(&free_queue, &cmd->node);
+	}
 
-    cmd = k_fifo_get(&used_queue, K_NO_WAIT);
+	node = sys_slist_get(&used_queue);
 
-    if (cmd == NULL) {
-        *line = NULL;
-        return 0;
-    }
+	if (node == NULL) {
+		cmd = NULL;
+		*line = NULL;
+		return 0;
+	}
 
-    *line = cmd->line;
-    return cmd->len;
+	cmd = CONTAINER_OF(node, struct line_input, node);
+	*line = cmd->line;
+	return cmd->len;
 }
 
 static int
 boot_uart_fifo_init(void)
 {
-    uart_dev = device_get_binding(CONFIG_UART_CONSOLE_ON_DEV_NAME);
-    u8_t c;
+	uart_dev = device_get_binding(CONFIG_UART_CONSOLE_ON_DEV_NAME);
+	u8_t c;
+
+	if (!uart_dev) {
+		return (-1);
+	}
 
-    if (!uart_dev) {
-        return (-1);
-    }
+	/* Enable all interrupts unconditionally. Note that this is due
+	 * to Zephyr issue #8393. This could should be removed once the
+	 * issue is fixed in upstream Zephyr. */
+	irq_unlock(0);
```

Comment:
@cvinayak can you please review this line?

Reply:
Any reason interrupts are enabled before the callback is setup? It is better to enable at function return.